Add index page for quizzes

cmkoller opened the pull request:
Add factories for easier testing of completed quizzes

Commit:
Add index page for quizzes
Add factories for easier testing of completed quizzes

Commit:
Display alert about incomplete quizzes

radavis commented on spec/features/student_views_quizzez_index_page_spec.rb:
```diff
@@ -0,0 +1,33 @@
+* [x] Display number of questions answered incorrectly per quiz.
+* [x] Link to the start of the quiz if it isn't complete.
+* [x] Link to the summary of the quiz if it is complete.
+* [] Provides a notice about the number of unfinished quizzes
```
Indent the text within the parenthesis.

radavis commented on spec/features/student_views_quizzez_index_page_spec.rb:
```diff
@@ -0,0 +1,33 @@
+    expect(page).to have_link "View Answers"
+    expect(page).to have_link "Complete Quiz"
+  end
+end
```
I think that the quizzes index page should be the first thing a student sees when they log in. Could you add this as a line in the acceptance criteria, as well as write a corresponding test?

radavis commented on app/views/quizzes/index.html.erb:
```diff
@@ -0,0 +1,39 @@
+<% if current_user.num_incomplete_quizzes > 0 %>
+  <% num = current_user.num_incomplete_quizzes %>
+  <div class="incomplete-alert">
+    You have <%= num %> incomplete quiz<%= "zes" if num > 1 %>!
```
This might be helpful: http://apidock.com/rails/ActionView/Helpers/TextHelper/pluralize

radavis commented on app/models/quiz.rb:
```diff
@@ -3,4 +3,19 @@ class Quiz < ActiveRecord::Base
+
+      !completed_answers.empty?
+    end
+  end
```
I would be a good idea to have a few model tests for this method. One for a user who has not answered any questions, one for a user who has answered some of the questions, and one for a user who has answered all the questions.

radavis commented on app/models/user.rb:
```diff
@@ -22,4 +22,17 @@ def self.find_or_create_via_omniauth(auth)
+
+  def num_incomplete_quizzes
+    Quiz.all.to_a.count { |quiz| !quiz.completed_by_student?(self) }
+  end
```
Model tests for these would be nice, too.

radavis commented:
I like to test the factories. Check out this from the factory_girl docs: https://github.com/thoughtbot/factory_girl/wiki/Testing-all-Factories-(with-RSpec)

radavis commented:
If there is a quiz in the database that the user has not yet taken, it should show up on the quizzes index page. 

In the future, we will assign quizzes through teams, but we are a few user stories away from that capability.

Commit:
Refactored and added test in response to comments
- Signing in now redirects to quiz index page
- Refactored incomplete quiz alert to use pluralize
- Added test for factories
- Renamed typo'd filename for user viewing quiz index page
- Added model tests for Quiz and User

radavis commented on app/views/quizzes/index.html.erb:
```diff
@@ -0,0 +1,39 @@
+  <div class="incomplete-alert">
+    You have <%= pluralize(num, "incomplete quiz")%>!
+  </div>
+<% end %>
```
One minor gripe: Could you change the method `num_incomplete_quizzes` to `incomplete_quizzes_count`? I might also switch a couple of lines to keep things DRY:

```
<% count = current_user.incomplete_quizzes_count %>
<% if count > 0 %>
  # ...
```

radavis commented on app/views/quizzes/index.html.erb:
```diff
@@ -0,0 +1,39 @@
+<h5>Total Correct Answers: <%= current_user.total_correct_questions %></h5>
+<h5>
+  Total Incorrect Answers:
+  <%= current_user.total_answered_questions - current_user.total_correct_questions %>
```
Could you move this logic to the User model?

radavis commented on spec/models/quiz_spec.rb:
```diff
@@ -5,4 +5,20 @@
+
+  it "returns a student's score" do
+    user = FactoryGirl.create(:user_with_answered_quiz)
+    quiz = Quiz.first
```
For this test, I think it would be better if you create a quiz with questions, and then create answer submissions for the user.

radavis commented on spec/models/quiz_spec.rb:
```diff
@@ -5,4 +5,20 @@
+
+    expect(quiz.completed_by_student?(user)).to eq(true)
+    expect(quiz.completed_by_student?(bad_user)).to eq(false)
+  end
```
Can you break this up into two separate tests? One for the good user and one for the bad user. Test that test one thing, and one thing only are easy to verify and easy to read.

radavis commented on app/assets/stylesheets/application.scss:
```diff
@@ -4,3 +4,9 @@
+  text-align: center;
+  background-color: tomato;
+  padding: .5em;
+}
```
Can you also max out the width of the table?

```
table {
  width: 100%;
}
```

Commit:
Updating in accordance to comments on PR
- Renamed "num_incomplete_quizzes" to "incomplete_quizzes_total"
- Defined method in User to return number of incorrect answers
- Refactored quizzes index page to DRY it up and move logic to model
- Deleted duplicated test file
- Refactored tests

Commit:
Tables are now full-width

radavis commented on spec/models/quiz_spec.rb:
```diff
@@ -5,4 +5,29 @@
+
+  it "returns true if it has been completed by a student" do
+    user = FactoryGirl.create(:user_with_answered_quiz)
+    quiz = Quiz.first
```
We can't guarantee that the first quiz is the one we want. Use the same strategy as in your user model for creating a quiz and answer submissions.

Commit:
Refactored one more test
